k8sapiserver/k8sapiserver.go

```diff
@@ -2,47 +2,42 @@ package k8sapiserver
 
 import (
 	"context"
-	"fmt"
+	"io/ioutil"
 	"net"
 	"net/http"
 	"net/http/httptest"
-	"time"
+	"os"
 
-	"github.com/google/uuid"
 	"golang.org/x/xerrors"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/apimachinery/pkg/util/wait"
-	authauthenticator "k8s.io/apiserver/pkg/authentication/authenticator"
-	"k8s.io/apiserver/pkg/authentication/authenticatorfactory"
-	authenticatorunion "k8s.io/apiserver/pkg/authentication/request/union"
 	"k8s.io/apiserver/pkg/authentication/user"
-	"k8s.io/apiserver/pkg/authorization/authorizerfactory"
-	authorizerunion "k8s.io/apiserver/pkg/authorization/union"
-	"k8s.io/apiserver/pkg/endpoints/openapi"
-	genericapiserver "k8s.io/apiserver/pkg/server"
-	"k8s.io/apiserver/pkg/server/options"
-	serverstorage "k8s.io/apiserver/pkg/server/storage"
-	"k8s.io/apiserver/pkg/storage/storagebackend"
-	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
-	"k8s.io/client-go/informers"
-	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes"
 	restclient "k8s.io/client-go/rest"
-	"k8s.io/component-base/version"
 	"k8s.io/klog/v2"
-	openapicommon "k8s.io/kube-openapi/pkg/common"
-	"k8s.io/kube-openapi/pkg/validation/spec"
-	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kube-aggregator/pkg/apiserver"
+	apiserverapp "k8s.io/kubernetes/cmd/kube-apiserver/app"
+	apiserverappopts "k8s.io/kubernetes/cmd/kube-apiserver/app/options"
 	"k8s.io/kubernetes/pkg/controlplane"
-	"k8s.io/kubernetes/pkg/kubeapiserver"
-	kubeletclient "k8s.io/kubernetes/pkg/kubelet/client"
-
-	generated "github.com/kubernetes-sigs/kube-scheduler-simulator/k8sapiserver/openapi"
+	authzmodes "k8s.io/kubernetes/pkg/kubeapiserver/authorizer/modes"
+	apiserveropts "k8s.io/kubernetes/pkg/kubeapiserver/options"
 )
 
-// StartAPIServer starts API server, and it make panic when a error happen.
+// This key is for testing purposes only and is not considered secure.
+const ecdsaPrivateKey = `-----BEGIN EC PRIVATE KEY-----
+MHcCAQEEIEZmTmUhuanLjPA2CLquXivuwBDHTt5XYwgIr/kA1LtRoAoGCCqGSM49
+AwEHoUQDQgAEH6cuzP8XuD5wal6wf9M6xDljTOPLX2i8uIp/C/ASqiIGUeeKQtX0
+/IR3qCXyThP/dbCiHrF3v1cuhBOHY8CLVg==
+-----END EC PRIVATE KEY-----`
+
+// StartAPIServer starts both the secure k8sAPIServer and proxy server to handle insecure serving, and it make panic when a error happen.
 func StartAPIServer(kubeAPIServerURL, etcdURL, frontendURL string) (*restclient.Config, func(), error) {
 	h := &APIServerHolder{Initialized: make(chan struct{})}
 	handler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.UserAgent() == "" {
+			req.Header.Set("User-Agent", restclient.DefaultKubernetesUserAgent())
+		}
```

Comment:
Sorry, could you please revert this? I'll try to fix this in another PR.

Reply:
reverted :)